qiskit_experiments/calibration_management/base_calibration_experiment.py

```diff
@@ -0,0 +1,340 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Base class for calibration-type experiments."""
+
+from abc import ABC
+from typing import Dict, Optional, Tuple
+
+from qiskit.providers.backend import Backend
+from qiskit.circuit import Parameter
+from qiskit.pulse import ScheduleBlock
+
+from qiskit_experiments.calibration_management.calibrations import Calibrations
+from qiskit_experiments.framework.base_experiment import BaseExperiment
+from qiskit_experiments.framework.experiment_data import ExperimentData
+from qiskit_experiments.exceptions import CalibrationError
+
+
+class BaseCalibrationExperiment(BaseExperiment, ABC):
+    """A mixin class to create calibration experiments.
+
+    This abstract class extends a characterization experiment by turning it into a
+    calibration experiment. Such experiments allow schedule management and updating of an
+    instance of :class:`Calibrations`. Furthermore, calibration experiments also specify
+    an auto_update variable which, by default, is set to True. If this variable,
+    is True then the run method of the experiment will call :meth:`block_for_results`
+    and update the calibrations instance once the backend has returned the data.
+
+    This mixin class inherits from the :class:`BaseExperiment` class since calibration
+    experiments by default call :meth:`block_for_results`. This ensures that the next
+    calibration experiment cannot proceed before the calibration parameters have been
+    updated. Developers that wish to create a calibration experiment must subclass this
+    base class and the characterization experiment. Therefore, developers that use this
+    mixin class must pay special attention to their class definition. Indeed, the first
+    class should be this mixin and the second class should be the characterization
+    experiment since the run method from the mixin must be used. For example, the rough
+    frequency calibration experiment is defined as
+
+    .. code-block:: python
+
+        RoughFrequency(BaseCalibrationExperiment, QubitSpectroscopy)
+
+    This ensures that the :meth:`run` method of :class:`RoughFrequency` will be the
+    run method of the :class:`BaseCalibrationExperiment` class. Furthermore, developers
+    must explicitly call the :meth:`__init__` methods of both parent classes.
```

Comment:
Regarding `run`, the inheritance order only matters if characterization experiments override `run` (which they generally shouldn't). Since `BaseCalibrationExperiment` inherits from `BaseExperiment`, its `run` will always be called before `BaseExperiment`'s. However, what we do with `__init__` could be more sensitive. Either `BaseCalibration.__init__` or the characterization experiment's `__init__` could be called first depending on the inheritance order.

Reply:
I think this should do the trick: 9f19a6e